### PR TITLE
[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference

### LINKED ISSUE #67244: Support automatic infer the ShardingSpec in init_from_local_shards
When using `init_from_local_shards` API to create sharded tensor, we set the ShardingSpec to be `EnumerableShardingSpec` for now, and it's created by using the global_sharded_tensor_metadata.metadatas. 

Ideally we want to infer the metadata list, and see if it's a ChunkShardingSpec (i.e. if all the metadata is split on a single dim?). If it is, we set it to ChunkShardingSpec instead, as that's the most popular use case, and we should be consistent when we create sharded tensor from scratch or from local shards.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### LINKED ISSUE #72022: [PT-D] To make the ShardedTensor Reshard more generic
### 🚀 The feature, motivation and pitch

We enabled the reshard for sharded tensor and added the partial tensor in https://github.com/pytorch/pytorch/pull/70079.
Below are several features we want to add:
1. To make `_reshard_output` API more generic and can handle the case when the output is Tuple.
2. To add padding for the case when there is uneven size of local tensor during the reshard of partial tensor.

The list goes on.

### Alternatives

_No response_

### Additional context

_No response_

### LINKED ISSUE #72092: [PT-D] More efficient ShardedTensor construction from local tensor.
### 🚀 The feature, motivation and pitch

We currently only have an API called `_init_from_local_shards` which actually performs an allgather: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/api.py#L351. We need a variation of this that doesn't do the validation for perf reasons or have a separate simpler API.

One proposal here is to have a new API like the following:

```
def from_local_tensor(tensor: torch.Tensor, *size: Sequence[int], spec: ShardingSpec) -> ShardedTensor
```
Each local rank can build the sharding metadata locally to avoid more expensive allgather. After implementing this, we then need to update the place where `_init_from_local_shards` is not necessary.

### Alternatives

_No response_

### Additional context

_No response_

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72733

To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

### CODE REVIEW

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 1]
```diff
@@ -357,15 +359,139 @@ def _init_from_local_shards(
...
+    def _init_from_local_tensor(
+        cls,
+        local_tensor: torch.Tensor,
+        sharding_spec: ShardingSpec,
+        *global_size: Sequence[int],
+        process_group: dist.ProcessGroup = None,
+        init_rrefs=False,
+    ) -> "ShardedTensor":
+        """
+        Initialize a ShardedTensor with only one local tensor with global sharded tensor
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/api.py:380]: nit: given global sharded tensor size?

File: torch/distributed/_shard/sharded_tensor/api.py [hunk 2]
```diff
@@ -357,15 +359,139 @@ def _init_from_local_shards(
...
+        size and sharding spec on each rank.
+
+        Args:
+            local_tensor (Tensor): Partial result stored in each rank.
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/api.py:384]: nit: local shard tensor stored in each rank? Partial results sounds a bit confusing

File: torch/distributed/_shard/sharding_spec/api.py [hunk 3]
```diff
@@ -110,3 +110,62 @@ def __post_init__(self):
...
+    """
+    placements = []
+    chunk_sharding_dim = None
+    chunk_offset_list = []
+    shard_size_list = []
+    # collect local shard metadatas from the global sharded_tensor_metadata
+    for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+        placements.append(shard_metadata.placement)
+        local_offsets = shard_metadata.shard_offsets
+        chunk_offset_list.append(sum(local_offsets))
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:136]: why we need to sum the local_offsets list?

File: torch/distributed/_shard/sharding_spec/api.py [hunk 4]
```diff
@@ -110,3 +110,62 @@ def __post_init__(self):
...
+        shard_size_list.append(shard_metadata.shard_sizes)
+        shard_dims = [idx for idx, e in enumerate(local_offsets) if e != 0]
+        if len(shard_dims) == 0:
+            continue
+        if len(shard_dims) != 1:
+            break
+        if not chunk_sharding_dim:
+            chunk_sharding_dim = shard_dims[0]
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:144]: is this saying that if we have `len(shard_dims) == 1` for a single metadata, we can assume this is a chunkshardingspec?

it might not always be the case i guess? i.e. a sharded tensor with global size: (10, 10)
Shard 0: offset: [0, 0], 
Shard 1: offset: [5, 0]
Shard 2: offset: [0, 5]
Shard 3: offset: [5, 5]

This is actually not a chunkShardingSpec but rather a EnumerableShardingSpec?

**Review by wanchaol — COMMENTED** [review]
Looks good overall, thanks for doing it! have some questions about the inference of the ChunkShardingSpec

File: torch/distributed/_shard/sharding_spec/api.py (hunk 4 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/api.py:144]: So, for this case. We will know it's an enum because 
For [5, 5], len(shard_dims) != 1.
For [5, 0] and then [0, 5], chunk_sharding_dim != shard_dims[0]. 

No matter what sequence you have in the sharding spec, we will know it's a enum sharding spec.

File: torch/distributed/_shard/sharding_spec/api.py (hunk 3 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/api.py:136]: Oh, because for here, we don't know the sharding col, so we just directly sum it. 

The assumption here is that, for chunk sharding case, the offset will always be [0, 0, ..., N, ..., 0]. So if we sum it, we  will get N regardless its sharding col.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 1 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/api.py:380]: Sure, let me update the comment here.

File: torch/distributed/_shard/sharded_tensor/api.py (hunk 2 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/api.py:384]: Maybe let's call it local tensor from the local shard in each rank?

**Comment by fduwjj** [comment]
Addressed the feedback from the reviewer.

**Comment by fduwjj** [comment]
Looks like the failed test is not relevant to this PR.

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py [hunk 5]
```diff
@@ -1781,6 +1785,124 @@ def test_with_rpc_names(self):
...
+    def _generate_st_from_chunk_local_tensor(self, st_size, sharding_spec):
+        chunk_sharding_dim = sharding_spec.dim
+        world_size = len(sharding_spec.placements)
+        split_size = get_split_size(st_size[chunk_sharding_dim], world_size)
+        shards_metadata = [None] * world_size
+        local_shard_meta = None
+        for idx, placement in enumerate(sharding_spec.placements):
+            shard_size = copy.deepcopy(st_size)
+            offsets = [0] * len(st_size)
+            offsets[chunk_sharding_dim] = min(
```
Comment (wanchaol) [inline test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:1798]: I guess we will have a easier way to generate the ShardedMetadatas for sharding spec when we have https://github.com/pytorch/pytorch/pull/72130

File: torch/distributed/_shard/sharded_tensor/_ops/linear.py [hunk 6]
```diff
@@ -218,8 +215,17 @@ def _handle_col_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
         result = torch.stack(results)  # type: ignore[arg-type]
     else:
         result = torch.cat(results)  # type: ignore[arg-type]
-    return _init_sharded_tensor_from_local_result(
-        weight, result, 0, -1, world_size, pg  # type: ignore[arg-type]
+    st_size = list(result.size())
+    st_size[-1] = weight.size(0)
+    new_sharding_spec = ChunkShardingSpec(
+        dim=-1,
+        placements=weight._sharding_spec.placements
```
Comment (wanchaol) [inline torch/distributed/_shard/sharded_tensor/_ops/linear.py:222]: nit: lets use the public API: weight.sharding_spec() instead.

File: torch/distributed/_shard/sharding_spec/api.py [hunk 7]
```diff
@@ -110,3 +110,65 @@ def __post_init__(self):
             rank = len(shard.shard_offsets)
 
         validate_non_overlapping_shards_metadata(self.shards)
+
+
+def _infer_sharding_spec_from_shards_metadata(shards_metadata):
+    """
+    Infer the sharding spec from the metadata of each shard of a ShardedTensor.
+    If the tensor is sharded only on one dimension, we then assume it's chunk sharded.
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:118]: nit: it's a ChunkShardingSpec?

File: torch/distributed/_shard/sharding_spec/api.py [hunk 8]
```diff
@@ -110,3 +110,65 @@ def __post_init__(self):
...
+        ]
+        chunk_spec = ChunkShardingSpec(
+            dim=chunk_sharding_dim,
+            placements=placements,
+        )
+        shard_sizes = [
+            x[chunk_sharding_dim]
+            for _, x in sorted(zip(chunk_offset_list, shard_size_list))
+        ]
+        if len(shard_sizes) == 1 or (
```
Comment (wanchaol) [inline torch/distributed/_shard/sharding_spec/api.py:163]: can you also make a comment to explain what this logic used for?

File: torch/distributed/_shard/sharding_spec/api.py (hunk 4 above)
Comment (wanchaol) [reply torch/distributed/_shard/sharding_spec/api.py:145]: I see, yeah make sense.

**Review by wanchaol — APPROVED** [review]
lgtm, thanks!

File: torch/distributed/_shard/sharded_tensor/_ops/linear.py (hunk 6 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharded_tensor/_ops/linear.py:222]: sure.

File: torch/distributed/_shard/sharding_spec/api.py (hunk 8 above)
Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/api.py:163]: sure, this is for corner cases like when enum sharding will also shard by one dimension but the offset is not the same, aka, the tensor is not evenly distributed.

File: test/distributed/_shard/sharded_tensor/test_sharded_tensor.py (hunk 5 above)
Comment (fduwjj) [reply test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:1798]: Yes, that's right. I will send a PR to rebase on your change. Let me add a TODO here.
